File: go/LeetCode/lastStoneWeightII.go
package main

func main() {
	
}

func lastStoneWeightII(stones []int) int {
	sum:=0
	for _,v :=range stones{
		sum+=v
	}
	cap:=sum/2
	dp := make([][]int, len(stones)+1)
	for i,_ := range dp{
		dp[i] = make([]int, cap+1)
	}

	for i:=1;i<=len(stones);i++ {
		for j:=0;j<=cap;j++ {
			if j>=stones[i-1] {
				dp[i][j] = max(dp[i-1][j], dp[i-1][j-stones[i-1]]+stones[i-1])
			}else {
				dp[i][j] = dp[i-1][j]
			}
		}
	}
	return sum-dp[len(stones)][cap]*2
}

func lastStoneWeightII1(stones []int) int {
	sum:=0
	for _,v := range stones{
		sum+=v
	}
	cap:=sum/2
	dp:=make([]int, cap+1)
	for _,num := range stones{
		for j:=cap;j>=num;j-- {
			dp[j] = max(dp[j-num]+num, dp[j])
		}
	}
	return sum-dp[cap]*2
}

File: python3/LeetCode/lastStoneWeightII.py
from typing import List


class Solution:
    def lastStoneWeightII(self, stones:List[int]) -> int:
        sum=0
        for v in stones:
            sum+=v
        cap = sum//2
        dp = [[0]*(cap+1) for i in range (len(stones)+1)]
        for i in range(1,len(stones)+1):
            for j in range(cap+1):
                if j>= stones[i-1]:
                    dp[i][j] = max(dp[i-1][j], dp[i-1][j-stones[i-1]]+stones[i-1])
                else:
                    dp[i][j] = dp[i-1][j]

        return sum-dp[len(stones)][cap]*2

    def lastStoneWeightII1(self, stones:List[int]) -> int:
        sum=0
        for v in stones:
            sum+=v
        cap = sum//2
        dp = [0]*(cap+1)
        for num in stones:
            for j in range(cap, num-1,-1):
                dp[j] = max(dp[j],dp[j-num]+num)
        return sum-dp[cap]*2


if __name__ == '__main__':
    stones = [2,7,4,1,8,1]
    print(Solution().lastStoneWeightII1(stones))